src/services/providers/provider-sources/load-balance-provider.ts

```diff
@@ -0,0 +1,138 @@
+import { ChainId, Timestamp, TimeString } from '@types';
+import { IProviderSource } from '../types';
+import { chainsUnion } from '@chains';
+import { createTransport, type Transport } from 'viem';
+import ms from 'ms';
+
+export type LoadBalanceSourceConfig = LoadBalanceConfig;
+export class LoadBalanceProviderSource implements IProviderSource {
+  constructor(private readonly sources: IProviderSource[], private readonly config: LoadBalanceConfig | undefined) {
+    if (sources.length === 0) throw new Error('Need at least one source to setup the provider source');
+  }
+
+  supportedChains() {
+    return chainsUnion(this.sources.map((source) => source.supportedChains()));
+  }
+
+  getViemTransport({ chainId }: { chainId: ChainId }) {
+    const transports = this.sources
+      .filter((source) => source.supportedChains().includes(chainId))
+      .map((source) => source.getViemTransport({ chainId }));
+    if (transports.length === 0) throw new Error(`Chain with id ${chainId} not supported`);
+    return loadBalance(transports, this.config);
+  }
+}
+
+function loadBalance(transports_: readonly Transport[], config: LoadBalanceConfig = {}): Transport {
+  const { minSuccessRate = 0.05, minSamples = 3, maxAttempts, samplesTtl = '30m' } = config;
+
+  const rpcMetrics: MethodMetrics[] = transports_.map(() => ({}));
+
+  return ({ chain, timeout, ...rest }) => {
+    const transports = transports_.map((t) => t({ chain, timeout, ...rest }));
+    return createTransport({
+      key: 'load-balance',
+      name: 'Load Balancing',
+      async request({ method, params }): Promise<any> {
+        let availableTransports = transports.map((transport, index) => ({ transport, index }));
+        let noAvailableTransportsError = undefined;
+        let attempts = 0;
+
+        while (!maxAttempts || attempts < maxAttempts) {
+          attempts++;
+          const filteredTransports = availableTransports
+            .map(({ transport, index }) => ({ transport, index, metrics: cleanupMetrics(rpcMetrics, index, method, samplesTtl) }))
+            .filter(({ metrics }) => {
+              return metrics.samples.length < minSamples || calculateSuccessRate(metrics) > minSuccessRate;
+            });
+          if (filteredTransports.length === 0) {
+            break; // No transports available
+          }
+
+          // Get the best transport
+          const { transport, index } = filteredTransports.reduce((best, current) => {
+            if (!current.metrics?.samples?.length && best.metrics?.samples?.length) return current;
+            if (!current.metrics.samples?.length || !best.metrics?.samples?.length) return best;
+            return calculateScore(current.metrics) > calculateScore(best.metrics) ? current : best;
```

Comment:
If think that if you change `calculateScore` so that it returns `Infinity` when there are no samples, then you can simply return `calculateScore(current.metrics) > calculateScore(best.metrics) ? current : best`